src/wifi_manager.c

```diff
@@ -616,83 +648,356 @@ wifi_manager_destroy()
+    }
+}
 
-    first_free = NULL;
+static void
+wifi_handle_ev_scan_done(void)
+{
+    ESP_LOGD(TAG, "MESSAGE: EVENT_SCAN_DONE");
+    /* As input param, it stores max AP number ap_records can hold. As output param, it receives the
+     * actual AP number this API returns.
+     * As a consequence, ap_num MUST be reset to MAX_AP_NUM at every scan */
+    ap_num = MAX_AP_NUM;
+    ESP_ERROR_CHECK(esp_wifi_scan_get_ap_records(&ap_num, accessp_records));
+    /* make sure the http server isn't trying to access the list while it gets refreshed */
+    if (wifi_manager_lock_json_buffer(pdMS_TO_TICKS(1000)))
+    {
+        /* Will remove the duplicate SSIDs from the list and update ap_num */
+        ap_num = ap_list_filter_unique(accessp_records, ap_num);
+        json_access_points_generate(accessp_records, ap_num);
+        wifi_manager_unlock_json_buffer();
+    }
+    else
+    {
+        ESP_LOGE(TAG, "could not get access to json mutex in wifi_scan");
+    }
+
+    http_server_start();
+    dns_server_start();
+}
 
-    for (int i = 0; i < *aps - 1; i++)
+static void
+wifi_handle_cmd_start_wifi_scan(void)
+{
+    ESP_LOGD(TAG, "MESSAGE: ORDER_START_WIFI_SCAN");
+
+    /* if a scan is already in progress this message is simply ignored thanks to the
+     * WIFI_MANAGER_SCAN_BIT uxBit */
+    const EventBits_t uxBits = xEventGroupGetBits(wifi_manager_event_group);
+    if (0 == (uxBits & (uint32_t)WIFI_MANAGER_SCAN_BIT))
     {
-        wifi_ap_record_t *ap = &aplist[i];
+        xEventGroupSetBits(wifi_manager_event_group, WIFI_MANAGER_SCAN_BIT);
+        /* wifi scanner config */
+        const wifi_scan_config_t scan_config = {
+            .ssid        = 0,
+            .bssid       = 0,
+            .channel     = 0,
+            .show_hidden = true,
+        };
+
+        const esp_err_t ret = esp_wifi_scan_start(&scan_config, false);
+        // sometimes when connecting to a network, a scan is started at the same time and then the scan
+        // will fail that's fine because we already have a network to connect and we don't need new scan
+        // results
+        // TODO: maybe fix this in the web page that scanning is stopped when connecting to a network
+        if (0 != ret)
+        {
+            ESP_LOGW(TAG, "scan start return: %d", ret);
+        }
+    }
+}
 
-        /* skip the previously removed APs */
-        if (ap->ssid[0] == 0)
-            continue;
+static void
+wifi_handle_cmd_load_sta(void)
+{
+    ESP_LOGI(TAG, "MESSAGE: ORDER_LOAD_AND_RESTORE_STA");
+    if (wifi_manager_fetch_wifi_sta_config())
+    {
+        ESP_LOGI(TAG, "Saved wifi found on startup. Will attempt to connect.");
+        wifiman_msg_send_cmd_connect_sta(CONNECTION_REQUEST_RESTORE_CONNECTION);
+    }
+    else
+    {
+        /* no wifi saved: start soft AP! This is what should happen during a first run */
+        ESP_LOGI(TAG, "No saved wifi found on startup. Starting access point.");
+        wifiman_msg_send_cmd_start_ap();
```

Comment:
This is a bit surprising, function load_sta ends up starting access point. Since this is an internal function it doesn't matter so much, up to you if you want to refactor it to e.g. return error code on no saved WiFi found and start the AP at the calling function.

Reply:
I don't want to change the algorithm much without unit-tests (the main purpose of this modifications is to fix codesmells).
And wifiman_msg_send_cmd_start_ap does not actually starts access point, it just sends message with a command to start access point.